### PR TITLE
reloader: Fixed reloader and added unit tests.

### PR DESCRIPTION
**Pull request by Bplotka**
Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: pkg/runutil/runutil.go [hunk 1]
```diff
@@ -22,18 +26,25 @@ func Repeat(interval time.Duration, stopc <-chan struct{}, f func() error) error
...
+func RetryWithLog(logger log.Logger, interval time.Duration, stopc <-chan struct{}, f func() error) error {
 	tick := time.NewTicker(interval)
 	defer tick.Stop()
 
 	var err error
 	for {
 		if err = f(); err == nil {
 			return nil
+		} else {
+			level.Error(logger).Log("msg", "function failed. Retrying","err", err)
```
Comment (fabxc) [inline pkg/runutil/runutil.go]: `go fmt` :P

File: pkg/reloader/reloader.go [hunk 2]
```diff
@@ -27,89 +28,153 @@ import (
...
+	initial := make(chan struct{}, 1)
+	if r.cfgFile != "" {
+		if err := configWatcher.Add(r.cfgFile); err != nil {
+			return errors.Wrap(err, "add config file watch")
+		}
+		level.Info(r.logger).Log(
+			"msg", "started watching config file for changes",
+			"in", r.cfgFile,
+			"out", r.cfgEnvsubstFile)
+		initial <- struct{}{}
```
Comment (fabxc) [inline pkg/reloader/reloader.go]: Mh, I'd prefer if we went back to not having all the logic in the for loop body but some methods instead. Then we can just call the function here once and then in the loop. Seems easier to grasp that using this `initial` channel here to trigger a first pass of all the code in the loop.

File: pkg/reloader/reloader.go [hunk 3]
```diff
@@ -27,89 +28,153 @@ import (
...
-	ruleDir          string
-	lastCfgHash      []byte
-	lastRulesHash    []byte
+	logger          log.Logger
+	reloadURL       *url.URL
+	cfgFile         string
+	cfgEnvsubstFile string
+	ruleDir         string
+	ruleInterval    time.Duration
+	retryInteval    time.Duration
```
Comment (fabxc) [inline pkg/reloader/reloader.go]: `r` missing

File: pkg/reloader/reloader.go (hunk 2 above)
Comment (Bplotka) [reply pkg/reloader/reloader.go]: mhm, sure I guess we can do so.

**Comment by Bplotka** [comment]
@fabxc  PTAL

**Comment by fabxc** [comment]
:+1:

**Pull request merged**